src/Bicep.Core/Semantics/Namespaces/DefaultNamespaceProvider.cs

```diff
@@ -5,29 +5,55 @@
 using Bicep.Core.Features;
+using Bicep.Core.Syntax;
 using Bicep.Core.TypeSystem;
 using Bicep.Core.TypeSystem.Az;
 
 namespace Bicep.Core.Semantics.Namespaces;
 
 public class DefaultNamespaceProvider : INamespaceProvider
 {
-    private delegate NamespaceType GetNamespaceDelegate(string aliasName, ResourceScope resourceScope, IFeatureProvider features);
+    private delegate NamespaceType GetNamespaceDelegate(
+        string aliasName,
+        ResourceScope resourceScope,
+        IFeatureProvider features,
+        ImportDeclarationSyntax? importDeclarationSyntax = null);
     private readonly ImmutableDictionary<string, GetNamespaceDelegate> providerLookup;
+    private readonly IAzResourceTypeLoaderFactory azResourceTypeLoaderFactory;
 
-    public DefaultNamespaceProvider(IAzResourceTypeLoader azResourceTypeLoader)
+    public DefaultNamespaceProvider(IAzResourceTypeLoaderFactory azResourceTypeLoaderFactory)
     {
-        var azResourceTypeProvider = new AzResourceTypeProvider(azResourceTypeLoader);
+        this.azResourceTypeLoaderFactory = azResourceTypeLoaderFactory;
         this.providerLookup = new Dictionary<string, GetNamespaceDelegate>
         {
-            [SystemNamespaceType.BuiltInName] = (alias, scope, features) => SystemNamespaceType.Create(alias, features),
-            [AzNamespaceType.BuiltInName] = (alias, scope, features) => AzNamespaceType.Create(alias, scope, azResourceTypeProvider),
-            [K8sNamespaceType.BuiltInName] = (alias, scope, features) => K8sNamespaceType.Create(alias),
+            [SystemNamespaceType.BuiltInName] = (alias, scope, features, ids) => SystemNamespaceType.Create(alias, features),
+            [AzNamespaceType.BuiltInName] = (alias, scope, features, ids) =>
+            {
+                var loader = azResourceTypeLoaderFactory.GetResourceTypeLoader(ids, features);
+                if (loader is null)
+                {
+                    return null!;
+                }
+                string providerVersion = "1.0.0"; //builtin version
+                if (features.DynamicTypeLoading)
+                {
+                    providerVersion = ids?.Specification.Version ?? providerVersion;
+                }
+                return AzNamespaceType.Create(alias, scope, new AzResourceTypeProvider(loader, providerVersion));
+
+            },
+            [K8sNamespaceType.BuiltInName] = (alias, scope, features, ids) => K8sNamespaceType.Create(alias),
         }.ToImmutableDictionary();
     }
 
-    public NamespaceType? TryGetNamespace(string providerName, string aliasName, ResourceScope resourceScope, IFeatureProvider features)
-        => providerLookup.TryGetValue(providerName)?.Invoke(aliasName, resourceScope, features);
+    public NamespaceType? TryGetNamespace(
+        string providerName,
+        string aliasName,
+        ResourceScope resourceScope,
+        IFeatureProvider features,
+        ImportDeclarationSyntax? importDeclarationSyntax = null)
```

Comment:
Actually wondering whether syntax tree concepts belong here as well, but less concerned than for type loaders.